### PR TITLE
Refactor UPnP exposed API to tf

### PR DESCRIPTION
**Pull request by DoSOfRedRiver**
https://rchain.atlassian.net/browse/CORE-1358

### CODE REVIEW

File: comm/src/main/scala/coop/rchain/comm/UPnP.scala [hunk 1]
```diff
@@ -35,78 +45,96 @@ object UPnP {
...
-            true
-          case _ =>
-            println(" [failed]")
-            false
-        }
-      }
+  private def printDevices[F[_]: Log: Monad](devices: UPnPDevices): F[Unit] =
+    for {
+      _ <- Log[F].info("INFO - No gateway devices found")
+      _ <- if (devices.all.isEmpty) Log[F].info("No need to open any port")
```
Comment (rabbitonweb) [inline comm/src/main/scala/coop/rchain/comm/UPnP.scala]: This is effectful. Wrap over `Task` and potentially use `ifM`

File: comm/src/main/scala/coop/rchain/comm/UPnP.scala [hunk 2]
```diff
@@ -35,78 +45,96 @@ object UPnP {
...
+
+      _ <- Log[F].info(showPortMappingHeader)
+      _ <- getPortMappings(gateway).toList.map(showPortMapping).traverse(Log[F].info)
+    } yield Some(gateway.getExternalIPAddress)
+
+  def assurePortForwarding[F[_]: Log: Sync](ports: List[Int]): F[Option[String]] =
+    for {
+      _       <- Log[F].info("trying to open ports using UPnP....")
+      devices <- Sync[F].delay(discover)
+      res <- if (devices.gateways.isEmpty) printDevices(devices) *> None.pure[F]
```
Comment (rabbitonweb) [inline comm/src/main/scala/coop/rchain/comm/UPnP.scala]: same as above

File: comm/src/main/scala/coop/rchain/comm/UPnP.scala [hunk 3]
```diff
@@ -35,78 +45,96 @@ object UPnP {
...
-            println(" [success]")
-            true
-          case _ =>
-            println(" [failed]")
-            false
-        }
-      }
+  private def printDevices[F[_]: Log: Monad](devices: UPnPDevices): F[Unit] =
+    for {
+      _ <- Log[F].info("INFO - No gateway devices found")
```
Comment (rabbitonweb) [inline comm/src/main/scala/coop/rchain/comm/UPnP.scala]: why method `printDevices` logs `"No gawaway devices found"`?

File: comm/src/main/scala/coop/rchain/comm/UPnP.scala [hunk 4]
```diff
@@ -35,78 +45,96 @@ object UPnP {
...
-  def discover: UPnPDevices = {
+  private def tryOpenPorts[F[_]: Log: Sync](
+      ports: List[Int],
+      devices: UPnPDevices
+  ): F[Option[String]] =
+    for {
+      _ <- Log[F].info(
+            s"Available gateway devices: ${devices.gateways.map(d => d.getFriendlyName).mkString(", ")}"
+          )
+      gateway = devices.validGateway.getOrElse(devices.gateways.head)
```
Comment (rabbitonweb) [inline comm/src/main/scala/coop/rchain/comm/UPnP.scala]: Isn't `validGateway` also effectful?

**Comment by rabbitonweb** [comment]
Overall good work! 👍 Please refer to calls to eager, effectful code

**Comment by DoSOfRedRiver** [comment]
@rabbitonweb 
```scala
devices <- Sync[F].delay(discover)
res <- if (devices.gateways.isEmpty) logGatewayEmpty(devices) *> None.pure[F]
            else tryOpenPorts(ports, devices)
```
Here we get `devices` with an effect, so returned structure is pure data, there is no need to enclose operations uses `UPnPDevices` over `F`.

**Comment by rabbitonweb** [comment]
@DoSOfRedRiver makes sense 👍

**Pull request merged**